### PR TITLE
feat(headless SSR): add hooks for ssr react commerce

### PR DESCRIPTION
**Pull request by y-lakhdar**
## TL;DR
Integration for server-side rendering with commerce functionalities. The changes include 
* Updates to the package exports
* New react hooks
* Creation of new types and hooks to facilitate SSR in commerce applications.

### Enhancements to SSR Support:
* Added a new export for `./ssr-commerce` in the `package.json` file. Implementer can now fetch all the goodies from the `@coveo/headless-react/ssr-commerce`

### New Utilities and Hooks:
* Created `useEngine` react hook.
* Added utility functions for building controller hooks (e.g. `useSearchBox`, `useFacetGenerator`), engine hooks, and state providers to manage SSR and CSR contexts.
* Added Static and Hydrated state providers for every solution types (search, listing, standalone).

> Note
> The controller hooks are programmatically generated based on the engine definition. You can find the logic in `buildControllerHooks`

### Type Definitions:
* Defined new types in `types.ts` to support the new SSR commerce functionalities, including `ReactCommerceEngineDefinition`, `ContextState`, and various hooks and state types.

https://coveord.atlassian.net/browse/KIT-3695

### CODE REVIEW

File: packages/headless-react/src/ssr-commerce/index.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export {defineCommerceEngine} from './commerce-engine.js';
+export type {ReactCommerceEngineDefinition} from './commerce-engine.js';
+export {MissingEngineProviderError} from '../errors.js';
```
Comment (alexprudhomme) [inline packages/headless-react/src/ssr-commerce/index.ts:3]: I see we do the same for headless-react/ssr but why are we exporting this error ? No users will ever want to import this and throw it.

Comment (alexprudhomme) [reply packages/headless-react/src/ssr-commerce/index.ts:3]: Right, I didn't think about catching it 😄

**Review by louis-bompart — APPROVED** [review]

File: packages/headless-react/src/ssr-commerce/common.tsx [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  >,
+  key: TKey
+): ControllerHook<InferControllerFromDefinition<TControllers[TKey]>> {
+  return () => {
+    const ctx = useContext(singletonContext.get());
+    if (ctx === null) {
+      throw new MissingEngineProviderError();
+    }
+
+    // Workaround to ensure that 'key' can be used as an index for 'ctx.controllers'. A more robust solution is needed.
```
Comment (louis-bompart) [inline packages/headless-react/src/ssr-commerce/common.tsx:54]: ??? Need a Jira?
